### DISCUSSION TITLE
How to auto-scroll to end of the chart?

### DISCUSSION THREAD

**Discussion by victory316** [patrykandpatrick/vico#334] [Questions]
Hello. I'm getting used to this useful library with compose.

Well, I'm showing the dynamic data with list that becomes larger(It keeps tracking the data), and it shows well with large size data.

But when new data updated to chart, it keeps scroll state in first, so it doesn't auto scroll when time goes by.

What option should I use to auto-scroll when new data added?

**Comment by patrickmichalik** [comment]
Hello! Please see `rememberChartScrollSpec`’s `initialScroll` and `autoScrollCondition` parameters.

**Reply by klaaslusterbroodvan** [reply]
I use those, but it still does not scroll to the end. This is difficult for beginners, I suppose. I have my nice graph, data gets added, but it does not scroll to the end:

```
setContent {
            SimpleChartTheme {
                // A surface container using the 'background' color from the theme
                Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {

                    val modelProducer = remember { ChartEntryModelProducer() }
                    val chartEntryModelProducer = remember { ChartEntryModelProducer() }
                    val mydata = remember { mutableStateListOf<FloatEntry>() }
                    var counter by remember { mutableStateOf(0f) }
                    val scrollState = rememberChartScrollState()
                    val scrollSpec = rememberChartScrollSpec( isScrollEnabled = true ,
                        autoScrollAnimationSpec = tween(0) ,
                        autoScrollCondition = AutoScrollCondition.OnModelSizeIncreased,
                        initialScroll = InitialScroll.End )

                    Column {

                        Chart(
                            chart = lineChart(),
                            chartModelProducer = chartEntryModelProducer,
                            startAxis = rememberStartAxis() ,
                            bottomAxis = rememberBottomAxis() ,
                            chartScrollState = scrollState ,
                            chartScrollSpec = scrollSpec
                        )

                        TextButton(
                            onClick = {
                                counter = counter + 1
                                val tmp = (5..10).random()
                                mydata.add( FloatEntry( counter , tmp.toFloat() ) )
                                chartEntryModelProducer.setEntries( mydata )

                            }
                        ) {
                            Text(text = "Click me")
                        }
                    }
                }
            }
        }
```

**Reply by Gowsky** [reply]
Hi @klaaslusterbroodvan, we’ve found a bug that prevents the auto scroll not to run. Thank you for the tip and the sample code. We will release a hotfix soon.

**Reply by klaaslusterbroodvan** [reply]
Thank you so much, that is greatly appreciated. The library is fantastic.

**Reply by klaaslusterbroodvan** [reply]
I don't want to open an issue for a related problem, namely that currently there is no documentation about how to programmatically set the initial value. The initial value can only be set to START or END. It is unclear how the position can be set. For example, I would like to have a button underneath my Chart and make it possible to "scroll" or "move" back 10.0f units. Is this done with the ScrollHandler, and if so, where can documentation be found?

**Reply by Gowsky** [reply]
You can use the `scrollState` from your example to scroll the chart by a desired value. See [`ChartScrollState#scroll`](https://patrykandpatrick.com/vico/api/compose/com.patrykandpatrick.vico.compose.chart.scroll/-chart-scroll-state/scroll.html).

Here’s an example:
````kotlin
val scope = rememberCoroutineScope()

// Inside of `onClick` lambda.
scope.launch {
    scrollState.scroll {
        scrollBy(10f)
    }
}
````

**Reply by patrickmichalik** [reply]
Please note that `scrollBy(10f)` scrolls the chart by 10 pixels, not by 10 times the distance between neighboring major entries (`xSpacing`), which is what I presume “10 units” was referring to. We plan on making the latter possible. Please see #265.

**Reply by patrickmichalik** [reply]
@klaaslusterbroodvan, [Vico 1.13.0 Alpha 2](https://github.com/patrykandpatrick/vico/releases/tag/v1.13.0-alpha.2) resolves the issue that you’ve reported. Cheers!

**Comment by klaaslusterbroodvan** [comment]
Thank you so much, I will try it tomorrow.--Sent from my Android phone with GMX Mail. Please excuse my brevity.On 24/09/2023, 18:06 Patrick Michalik ***@***.***> wrote:

   
  @klaaslusterbroodvan, Vico 1.13.0 Alpha 2 resolves the issue that you’ve reported. Cheers!

**Reply by patrickmichalik** [reply]
You’re welcome!
